File: UIViewControllerTransitions/Classes/AnimatedTransitioning.m
//
//  AnimatedTransitioning.m
//  UIViewControllerTransitions
//
//  Created by Steve Kim on 5/12/16.
//  Modified by Steve Kim on 4/14/17.
//      - Renew design and add new feature interactive transition
//  Modified by Steve Kim on 8/14/17.
//      - Refactoring design for 3.0.0
//

#import "AnimatedTransitioning.h"
#import "UIViewControllerTransition.h"

@implementation AnimatedTransitioning

#pragma mark - Overridden: AbstractAnimatedTransitioning

- (UIViewController *)aboveViewController {
    return _presenting ? self.toViewController : self.fromViewController;
}

- (UIViewController *)belowViewController {
    return _presenting ? self.fromViewController : self.toViewController;
}

- (void)animateTransition:(id <UIViewControllerContextTransitioning>)transitionContext {
    [super animateTransition:transitionContext];
    
    self.fromViewController.modalPresentationCapturesStatusBarAppearance = YES;
    self.toViewController.modalPresentationCapturesStatusBarAppearance = YES;
    
    if (!transitionContext.isInteractive) {
        [self.fromViewController beginAppearanceTransition:NO animated:transitionContext.isAnimated];
        [self.toViewController beginAppearanceTransition:YES animated:transitionContext.isAnimated];
    }
    
    if (_presenting) {
        [self animateTransitionForPresenting:transitionContext];
    } else {
        [self animateTransitionForDismission:transitionContext];
    }
}

- (void)clear {
    [super clear];
    
    [self.fromViewController beginAppearanceTransition:YES animated:NO];
    [self.toViewController beginAppearanceTransition:NO animated:NO];
    [self.toViewController.view removeFromSuperview];
    [self.fromViewController endAppearanceTransition];
    [self.toViewController endAppearanceTransition];
}

#pragma mark - Protected methods

- (void)animateTransitionForDismission:(id<UIViewControllerContextTransitioning>)transitionContext {
}

- (void)animateTransitionForPresenting:(id<UIViewControllerContextTransitioning>)transitionContext {
}

@end

File: AnimatedTransitionKit/Classes/Transitions/Base/AnimatedTransitioning.swift
//
//  AnimatedTransitioning.swift
//  Pods
//
//  Created by Minwoo on 9/4/25.
//

import UIKit


@objcMembers open class AnimatedTransitioning: AbstractAnimatedTransitioning {
    private var isDismissTransitionInitialized = false
    open var isPresenting: Bool = false   // Objective-C 호환성으로 public 사용

    // MARK: - Overridden: AbstractAnimatedTransitioning

    override public var aboveViewController: UIViewController? {
        return isPresenting ? toViewController : fromViewController
    }

    override public var belowViewController: UIViewController? {
        return isPresenting ? fromViewController : toViewController
    }

    override public func animateTransition(using transitionContext: UIViewControllerContextTransitioning) {
        super.animateTransition(using: transitionContext)

        isDismissTransitionInitialized = !isPresenting

        fromViewController?.modalPresentationCapturesStatusBarAppearance = true
        toViewController?.modalPresentationCapturesStatusBarAppearance = true

        if isAllowsAppearanceTransition && !transitionContext.isInteractive {
            belowViewController?.beginAppearanceTransition(!isPresenting, animated: transitionContext.isAnimated)
        }

        if isPresenting {
            animateTransition(forPresenting: transitionContext)
        } else {
            animateTransition(forDismission: transitionContext)
        }
    }

    override public func clear() {
        if isDismissTransitionInitialized { return }

        if isAllowsDeactivating {
            belowViewController?.view.alpha = 1
            belowViewController?.view.transform = .identity
            belowViewController?.view.tintAdjustmentMode = .normal
            belowViewController?.view.isHidden = false
        }

        if isAllowsAppearanceTransition {
            aboveViewController?.beginAppearanceTransition(false, animated: false)
            belowViewController?.beginAppearanceTransition(true, animated: false)
        }
        
        aboveViewController?.view.removeFromSuperview()

        if isAllowsAppearanceTransition {
            aboveViewController?.endAppearanceTransition()
            belowViewController?.endAppearanceTransition()
        }
    }

    open func animateTransition(forDismission transitionContext: UIViewControllerContextTransitioning) {
        // To be overridden by subclasses
    }
    
    open func animateTransition(forPresenting transitionContext: any UIViewControllerContextTransitioning) {
        toViewController?.view.frame = fromViewController?.view.bounds ?? .zero
    }
}
